src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -310,35 +411,27 @@ public static ValuesNode create() {
 
                 if (start > -1 && end > -1) {
                     Rope rope = makeSubstringNode.executeMake(sourceRope, start, end - start);
-                    values[n] = createSubstring(source, rope);
+                    DynamicObject string = allocateNode.allocate(Layouts.BASIC_OBJECT.getLogicalClass(source), Layouts.STRING.build(false, false, rope));
+                    values[n] = string;
```

Comment:
These Strings should be tainted based on whether the MatchData is taitned I think.